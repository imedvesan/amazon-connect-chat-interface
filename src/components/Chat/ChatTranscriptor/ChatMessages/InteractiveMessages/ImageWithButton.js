// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
 
import React, { useState } from "react";
import styled from "styled-components";
import PT from "prop-types";
import { Button } from "connect-core";
import {
  ReactiveImage,
  TextSection,
  ResponsesSection,
  Title,
  Subtitle
} from "../InteractiveMessage";
 
// have to calculate max height to maintain 16:9 aspect ratio
const ImageContainer = styled.div`
  max-height: calc(95vw * (9/16));
  overflow: hidden;
  display: ${props => props.showImage ? "flex" : "none"};
`;
 
const ButtonComponent = styled(Button)`
  display: flex;
  justify-content: center;
  width: 100%;
  max-width: none;
  border: ${({ theme }) => theme.globals.baseBorder};  
  white-space: pre-line;
 
  &:hover {
    background: ${({ theme }) => theme.color.primary};
  }
  
  &:last-of-type {
    border-bottom-left-radius: ${({ theme }) => theme.spacing.mini};
    border-bottom-right-radius: ${({ theme }) => theme.spacing.mini};
    margin-bottom: 0;   
  }
  
`;
 
function ButtonElement({ element, onClick }) {
  const { title, subtitle } = element;
 
  return (
    <ButtonComponent value={title}  onClick={onClick}>
      <div>
        <Title>{title}</Title>
        {subtitle && <Subtitle>{subtitle}</Subtitle>}
      </div>
    </ButtonComponent>
  );
}
 
ImageWithButton.propTypes = {
  content: PT.object.isRequired,
  addMessage: PT.func.isRequired
};
 
export default function ImageWithButton({ content, addMessage }) {
    console.error(content);
  // assumptions: version 1, image data is URL. Guarenteed title exists, at least 1 element.
  const { title, subtitle, elements, imageData, imageDescription } = content;
  const [ imageLoaded, setImageLoaded ] = useState(false);
  
  function onImageLoad(){
    setImageLoaded(true);
  }
 
  function onItemClick(e) {
    addMessage({ text: e.currentTarget.value });
  }
 
  console.log("Inside ImageWithButton");

  return (
    <>
      {(imageData) &&
      <ImageContainer showImage={imageLoaded}>
        <ReactiveImage imageSrc={imageData} imageDescription={imageDescription} onImageLoad={onImageLoad}/>
      </ImageContainer>
      }      
      
      <TextSection>
      <Title>{title}</Title>
      {subtitle && <Subtitle>{subtitle}</Subtitle>}
    </TextSection>
    <ResponsesSection>
      <div>
        {elements.map((button, index) =>
          <ButtonElement key={'element-' + index} element={button} onClick={onItemClick} />
        )}
      </div>
    </ResponsesSection>
    </>
  );
};