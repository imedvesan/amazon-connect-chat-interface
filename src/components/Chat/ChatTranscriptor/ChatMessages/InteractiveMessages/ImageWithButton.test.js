// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import React from 'react';
import { render, fireEvent } from "@testing-library/react"
import { InteractiveMessageType } from "../../../datamodel/Model";
import { ThemeProvider } from "../../../../../theme";
import ImageWithButton from "./ImageWithButton";
 
const mockImageWithButtonContent = {
    title: "MessageTitle",
    subtitle: "MessageSubTitle",
    elements: [
        {
            title: "Button1Title",
        },
        {
            title: "Button2Title",
        }
    ],
    imageData: "ImageData",
    imageDescription: "ImageDescription"
};


const mockImageWithButtonNoImageContent = {
    title: "MessageTitle",
    subtitle: "MessageSubTitle",
    elements: [
        {
            title: "Button1Title",
        },
        {
            title: "Button2Title",
        }
    ],
};
 
const mockImageWithButtonNoButtonContent = {
    title: "MessageTitle",
    subtitle: "MessageSubTitle",
    elements: [],
    imageData: "ImageData",
    imageDescription: "ImageDescription"
};

 
let mockImageWithButton;
let mockProps;
 
function renderElement(props) {
    mockImageWithButton = render(
        <ThemeProvider>
            <ImageWithButton {...props}/>
        </ThemeProvider>
    );
}
 
test("Style should match the snapshot", () => {
    const addMessage = jest.fn().mockResolvedValue(undefined);
    mockProps = {content: mockImageWithButtonContent, addMessage: addMessage};

    renderElement(mockProps);
    expect(mockImageWithButton).toMatchSnapshot();
});
 
test("Should be able to use ImageWithButton", () => {
    const addMessage = jest.fn().mockResolvedValue(undefined);
    mockProps = {content: mockImageWithButtonContent, addMessage: addMessage};

    renderElement(mockProps);
 
    expect(mockImageWithButton.getByText("MessageTitle")).toBeDefined();
    expect(mockImageWithButton.getByAltText("ImageDescription")).toBeDefined();
 
    expect(mockImageWithButton.getByText("Button1Title")).toBeDefined();
    expect(mockImageWithButton.getByText("Button2Title")).toBeDefined();
 
    fireEvent.click(mockImageWithButton.getByText("Button1Title"));
    expect(mockProps.addMessage).toHaveBeenCalledTimes(1);
    expect(mockProps.addMessage).toHaveBeenCalledWith({text: "Button1Title"});
});

test("should render without image if no image data passed in", () => {
    const addMessage = jest.fn().mockResolvedValue(undefined);
    mockProps = {content: mockImageWithButtonNoImageContent, addMessage: addMessage};

    renderElement(mockProps);
 
    expect(mockImageWithButton.getByText("MessageTitle")).toBeDefined();
    expect(mockImageWithButton.queryByText("ImageDescription")).toBeNull();
 
    expect(mockImageWithButton.getByText("Button1Title")).toBeDefined();
    expect(mockImageWithButton.getByText("Button2Title")).toBeDefined();
 
    fireEvent.click(mockImageWithButton.getByText("Button1Title"));
    expect(mockProps.addMessage).toHaveBeenCalledTimes(1);
    expect(mockProps.addMessage).toHaveBeenCalledWith({text: "Button1Title"});
});

test("should render without buttons if no button elements passed in", () => {
    const addMessage = jest.fn().mockResolvedValue(undefined);
    mockProps = {content: mockImageWithButtonNoButtonContent, addMessage: addMessage};

    renderElement(mockProps);
 
    expect(mockImageWithButton.getByText("MessageTitle")).toBeDefined();
    expect(mockImageWithButton.getByAltText("ImageDescription")).toBeDefined();
 
    expect(mockImageWithButton.queryByText("Button1Title")).toBeNull();
    expect(mockImageWithButton.queryByText("Button2Title")).toBeNull();
});