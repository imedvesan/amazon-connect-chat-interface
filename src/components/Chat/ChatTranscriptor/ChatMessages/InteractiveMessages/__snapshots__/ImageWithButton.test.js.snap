// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Style should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c1 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  color: inherit;
  margin: 0;
}

.c1:last-child {
  margin: 0;
}

.c7 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  white-space: nowrap;
  color: #111111;
  cursor: pointer;
  text-align: center;
  vertical-align: middle;
  border-width: 0px;
  border-style: solid;
  padding-right: 10px;
  padding-left: 10px;
  font-family: Arial,Helvetica,"sans-serif";
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-width: 260px;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  color: undefined;
  padding-top: 0.45rem;
  padding-bottom: 0.45rem;
  font-size: 0.875rem;
  box-shadow: 1px 2px 3px 0px rgba(0,0,0,0.1);
}

.c7:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c7:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c7:active {
  border-color: #fff #fff #fff;
}

.c7:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c7:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c7 > img,
.c7 .loader {
  margin-right: undefined;
  position: relative;
  top: 1px;
}

.c7:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c7:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c7:active {
  border-color: #fff #fff #fff;
}

.c7:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c7:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c2 {
  font-weight: bold !important;
}

.c5 {
  color: #555555;
}

.c4 {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.c0 {
  padding: 14px;
  text-align: left;
}

.c6 {
  position: relative;
  white-space: pre-line;
  border-radius: 6px;
}

.c3 {
  max-height: calc(95vw * (9/16));
  overflow: hidden;
  display: none;
}

.c8 {
  display: inline;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0 10px 10px 0;
  max-width: none;
  border: 1px solid #DDDDDD;
  white-space: pre-line;
  border-radius: 6px;
}

.c8:hover {
  background: #06adcb;
}

<body>
    <div>
      <div
        class="c0"
      >
        <p
          class="c1 c2"
        >
          MessageTitle
        </p>
      </div>
      <div
        class="c3"
      >
        <img
          alt="ImageDescription"
          class="c4"
          src="ImageData"
        />
      </div>
      <div
        class="c0"
      >
        <p
          class="c1 c5"
        >
          MessageSubTitle
        </p>
      </div>
      <div
        class="c6"
      >
        <div>
          <button
            class="c7 c8"
            data-selected="false"
            type="default"
            value="Button1Title"
          >
            <div>
              <p
                class="c1 c2"
              >
                Button1Title
              </p>
            </div>
          </button>
          <button
            class="c7 c8"
            data-selected="false"
            type="default"
            value="Button2Title"
          >
            <div>
              <p
                class="c1 c2"
              >
                Button2Title
              </p>
            </div>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-iRbamj hpGXcd"
    >
      <p
        class="sc-bZQynM fraGhu sc-jAaTju hYWMdF"
      >
        MessageTitle
      </p>
    </div>
    <div
      class="sc-gipzik inmHgs"
    >
      <img
        alt="ImageDescription"
        class="sc-gPEVay cbWEyG"
        src="ImageData"
      />
    </div>
    <div
      class="sc-iRbamj hpGXcd"
    >
      <p
        class="sc-bZQynM fraGhu sc-jDwBTQ jVYcAQ"
      >
        MessageSubTitle
      </p>
    </div>
    <div
      class="sc-jlyJG fUHAZN"
    >
      <div>
        <button
          class="sc-gzVnrw fgSGc sc-csuQGl gdWUTH"
          data-selected="false"
          type="default"
          value="Button1Title"
        >
          <div>
            <p
              class="sc-bZQynM fraGhu sc-jAaTju hYWMdF"
            >
              Button1Title
            </p>
          </div>
        </button>
        <button
          class="sc-gzVnrw fgSGc sc-csuQGl gdWUTH"
          data-selected="false"
          type="default"
          value="Button2Title"
        >
          <div>
            <p
              class="sc-bZQynM fraGhu sc-jAaTju hYWMdF"
            >
              Button2Title
            </p>
          </div>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
